### PR TITLE
Enhance inference in LieAlgebras, GSets, and MatrixGroups

### PR DESCRIPTION
**Pull request by lgoettgens**
Most of this was found using JET.

This needs https://github.com/Nemocas/AbstractAlgebra.jl/pull/1864 to be available.

### CODE REVIEW

File: src/Groups/matrices/MatGrp.jl [hunk 1]
```diff
@@ -464,7 +465,7 @@ Return the underlying matrix of `x`.
 """
 function matrix(x::MatrixGroupElem)
   if !isdefined(x, :elm)
-    x.elm = preimage_matrix(_ring_iso(x.parent), x.X)
+    x.elm = preimage_matrix(_ring_iso(parent(x)), x.X)
```
Comment (ThomasBreuer) [inline src/Groups/matrices/MatGrp.jl]:
```suggestion
    x.elm = preimage_matrix(_ring_iso(parent(x)), GapObj(x))
```

File: src/Groups/matrices/MatGrp.jl [hunk 2]
```diff
@@ -1034,7 +1035,7 @@ function Base.:^(H::MatrixGroup, y::MatrixGroupElem)
 end
 
 function Base.rand(rng::Random.AbstractRNG, C::GroupConjClass{S,T}) where S<:MatrixGroup where T<:MatrixGroup
-   H = matrix_group(C.X.ring, C.X.deg)
+   H = matrix_group(base_ring(C.X), degree(C.X))
```
Comment (ThomasBreuer) [inline src/Groups/matrices/MatGrp.jl]:
```suggestion
   H = matrix_group(base_ring(acting_group(C)), degree(acting_group(C)))
```

File: experimental/LieAlgebras/src/AbstractLieAlgebra.jl [hunk 3]
```diff
@@ -218,9 +218,15 @@ end
 function lie_algebra(
   basis::Vector{AbstractLieAlgebraElem{C}}; check::Bool=true
 ) where {C<:FieldElem}
-  parent_L = parent(basis[1])
-  @req all(parent(x) === parent_L for x in basis) "Elements not compatible."
-  R = coefficient_ring(parent_L)
+  @req !isempty(basis) "Basis must not be empty, or provide the lie algebra as first argument"
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/AbstractLieAlgebra.jl]:
```suggestion
  @req !isempty(basis) "Basis must not be empty, or provide the Lie algebra as first argument"
```

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl [hunk 4]
```diff
@@ -202,12 +202,17 @@ end
 function lie_algebra(
   basis::Vector{LinearLieAlgebraElem{C}}; check::Bool=true
 ) where {C<:FieldElem}
-  parent_L = parent(basis[1])
-  @req all(parent(x) === parent_L for x in basis) "Elements not compatible."
-  R = coefficient_ring(parent_L)
-  n = parent_L.n
+  @req !isempty(basis) "Basis must not be empty, or provide the lie algebra as first argument"
```
Comment (ThomasBreuer) [inline experimental/LieAlgebras/src/LinearLieAlgebra.jl]:
```suggestion
  @req !isempty(basis) "Basis must not be empty, or provide the Lie algebra as first argument"
```

**Review by ThomasBreuer — APPROVED** [review]
Nice.
I have marked just a few places where minor improvements are possible.

**Comment by ThomasBreuer** [comment]
Strange: On my computer, I cannot reproduce the errors reported for `test/Groups/matrixgroups.jl` in "test (short)".

File: experimental/LieAlgebras/src/LieAlgebraIdeal.jl [hunk 5]
```diff
@@ -99,21 +99,21 @@ function (I::LieAlgebraIdeal)()
...
 end
 
 @doc raw"""
-    (I::LieAlgebraIdeal{C})(v::Vector{C}) -> LieAlgebraElem{C}
+    (I::LieAlgebraIdeal{C})(v::AbstractVector{C}) -> LieAlgebraElem{C}
 
 Return the element of `I` with coefficient vector `v`.
 """
-function (I::LieAlgebraIdeal{C})(v::Vector{C}) where {C<:FieldElem}
+function (I::LieAlgebraIdeal{C})(v::AbstractVector{C}) where {C<:FieldElem}
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/LieAlgebraIdeal.jl:116]: At first I was concerned that this change might lead to funky behaviour if `v` does not use 1-based indexing. But then I noticed the call `matrix(coefficient_ring(I), 1, length(v), v)` below, so I think that will throw a suitable error in that case. Fine by me then.

**Review by fingolfin — APPROVED** [review]
Look sensible to me.

File: experimental/LieAlgebras/src/LieAlgebraIdeal.jl (hunk 5 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LieAlgebraIdeal.jl:116]: This gets caught by https://github.com/Nemocas/AbstractAlgebra.jl/pull/1864/files#diff-77fc321e067710edc4d1df4198d7b3c983ee423acee3295a333ac68533656322R1076, so it should be fine

**Pull request merged**